src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java

```diff
@@ -174,23 +250,54 @@ public ClientT client() {
+                        public Completed<RequestT, ResponseT, ClientT, ModelT, CallbackT> retryErrorFilter(final Callback<
+                            ? super RequestT, Exception, ClientT, ModelT, CallbackT, Boolean> retryFilter) {
+                            return handleError(((request, exception, client_, model_, context_) -> {
+                                if (retryFilter.invoke(request, exception, client_, model_, context_)) {
+                                    throw RetryableException.builder().build();
                                 }
-                                return defaultHandler(request, exception, client1, model1, context1);
-                            });
+                                return defaultHandler(request, exception, client_, model_, context_);
+                            }));
                         }
 
                         @Override
-                        public Completed<RequestT, ResponseT, ClientT, ModelT, CallbackT> exceptHandler(Callback<? super RequestT,
-                            Exception, ClientT, ModelT, CallbackT, ProgressEvent<ModelT, CallbackT>> handler) {
-                            this.exceptHandler = handler;
+                        public Completed<RequestT, ResponseT, ClientT, ModelT, CallbackT> handleError(ExceptionPropagate<
+                            ? super RequestT, Exception, ClientT, ModelT, CallbackT, ProgressEvent<ModelT, CallbackT>> handler) {
+                            getExceptionHandler(handler);
                             return this;
                         }
 
+                        private
+                            Callback<? super RequestT, Exception, ClientT, ModelT, CallbackT, ProgressEvent<ModelT, CallbackT>>
+                            getExceptionHandler(final ExceptionPropagate<? super RequestT, Exception, ClientT, ModelT, CallbackT,
+                                ProgressEvent<ModelT, CallbackT>> handler) {
+                            if (this.exceptHandler == null) {
+                                this.exceptHandler = ((request, exception, client_, model_, context_) -> {
+                                    ProgressEvent<ModelT, CallbackT> event = null;
+                                    Exception ex = exception;
+                                    ExceptionPropagate<? super RequestT, Exception, ClientT, ModelT, CallbackT,
+                                        ProgressEvent<ModelT, CallbackT>> inner = handler;
+                                    boolean defaultHandler = false;
+                                    do {
+                                        try {
+                                            event = inner.invoke(request, ex, client_, model_, context_);
+                                        } catch (RetryableException e) {
+                                            break;
+                                        } catch (Exception e) {
+                                            if (defaultHandler) {
```

Comment:
so if we get a non `RetryableException` then we give execution another chance and if same scenario hits again then we throw `FRAMEWORK ERROR..`? if so why is it a framework error?

Reply:
The default handler always returns ProgressEvent or RetryableException. If we get an exception when delegate to default handler, it will loop forever. Hence we break out.